src/state/token_index.ml

```diff
@@ -0,0 +1,601 @@
+      Lwt_io.with_file ~mode:Lwt_io.Output filename @@
+        fun channel ->
+        Lwt_io.write channel (serialise data)
+end
+
+(* inspired from learnocaml_data.ml *)
+let enc_check_version_1 file enc =
+  J.conv
+    (fun data -> ("1", data))
+    (fun (version, data) ->
+       begin
+         match version with
+         | "1" -> ()
+         | _ ->
+             let msg = Format.asprintf "%s: unknown version %s" file version in
+             raise (J.Cannot_destruct ([], Failure msg))
+       end ;
+       data)
+    (J.merge_objs (J.obj1 (J.req "learnocaml_version" J.string))
+       (J.obj1 (J.req file enc)))
+
+module BaseTokenIndex (RW: IndexRW) = struct
+  let rw = RW.init ()
+  let file = "token.json"
+
+  let enc = enc_check_version_1 file @@ J.list Token.enc
+
+  let parse = Json_codec.decode enc
+  let serialise_str = Json_codec.encode ~minify:false
+                        (enc_check_version_1 file J.(list string))
+  let serialise = Json_codec.encode ~minify:false enc
+
+  let create_index sync_dir =
+    let found_indexes =
+      let rec scan f d acc =
+        let rec aux s acc =
+          Lwt.catch (fun () ->
+              Lwt_stream.get s >>= function
+              | Some ("." | ".." | "data") -> aux s acc
+              | Some x -> scan f (d / x) acc >>= aux s
+              | None -> Lwt.return acc)
+          @@ function
+          | Unix.Unix_error (Unix.ENOTDIR, _, _) -> f d acc
+          | Unix.Unix_error _ -> Lwt.return acc
+          | e -> Lwt.fail e
+        in
+        aux (Lwt_unix.files_of_directory (sync_dir / d)) acc
+      in
+      scan (fun d acc ->
+          let d =
+            if Filename.basename d = "save.json" then Filename.dirname d
+            else d
+          in
+          let stok = String.map (function '/' | '\\' -> '-' | c -> c) d in
+          if Token.check stok then
+            Lwt.return (stok :: acc)
+          else
+            Lwt.return acc
+        ) "" [] in
+    Lwt_io.printl "[INFO] Regenerating the token index..." >>= fun () ->
+    found_indexes >>= RW.write rw (sync_dir / indexes_subdir / file) serialise_str
+
+  let get_file sync_dir name =
+    let filename = (sync_dir / indexes_subdir / name) in
+    let create () =
+          create_index sync_dir >>= fun () ->
+          RW.read rw filename parse in
+    if Sys.file_exists filename then
+      Lwt.catch
+        (fun () -> RW.read rw filename parse)
+        (fun _exn ->
+           (* Note: this error handler may be adapted later to be more conservative?
+              it does not matter now as sync/data/token.json is not a critical file, and
+              can be regenerated. *)
+           create ())
+    else
+      create ()
+
+  let get_tokens sync_dir =
+    get_file sync_dir file
+
+  let add_token sync_dir token =
+    get_tokens sync_dir >>= fun tokens ->
+    if not (List.exists (fun found_token -> found_token = token) tokens) then
```

Comment:
```suggestion
    if not (List.mem token tokens) then
```

Please see my review at https://github.com/ocaml-sf/learn-ocaml/pull/354#pullrequestreview-716905707 , I believe it still applies here.